cirq/ops/pauli_string.py

```diff
@@ -1008,63 +1054,292 @@ def _from_json_dict_(  # type: ignore
+        $C P C^\dagger$.
+
+        Args:
+            ops: A stabilizer operation or nested collection of stabilizer
+                operations.
+
+        Returns:
+            The mutable pauli string that was mutated.
+        """
+        for clifford in op_tree.flatten_to_ops(ops):
+            for op in _decompose_into_cliffords(clifford):
+                ps = [self.pauli_int_dict.pop(q, 0) for q in op.qubits]
+                if not any(ps):
+                    continue
+                gate = op.gate
+
+                if isinstance(gate, clifford_gate.SingleQubitCliffordGate):
+                    out = gate.transform(cast(cirq.Pauli, _INT_TO_PAULI[ps[0]]))
+                    if out.flip:
+                        self.coefficient *= -1
+                    self.pauli_int_dict[
+                        op.qubits[0]] = PAULI_GATE_LIKE_TO_INDEX_MAP[out.to]
+
+                elif isinstance(gate,
+                                pauli_interaction_gate.PauliInteractionGate):
+                    q0, q1 = op.qubits
+                    p0 = _INT_TO_PAULI[ps[0]]
+                    p1 = _INT_TO_PAULI[ps[1]]
+
+                    # Kick across Paulis that anti-commute with the controls.
+                    kickback_0_to_1 = not protocols.commutes(p0, gate.pauli0)
+                    kickback_1_to_0 = not protocols.commutes(p1, gate.pauli1)
+                    kick0 = gate.pauli1 if kickback_0_to_1 else identity.I
+                    kick1 = gate.pauli0 if kickback_1_to_0 else identity.I
+                    self.__imul__({q0: p0, q1: kick0})
+                    self.__imul__({q0: kick1, q1: p1})
+
+                    # Decompose inverted controls into single-qubit operations.
+                    if gate.invert0:
+                        self.inplace_after(gate.pauli1(q1))
+                    if gate.invert1:
+                        self.inplace_after(gate.pauli0(q0))
+
+                else:
+                    # coverage: ignore
+                    raise NotImplementedError(
+                        f"Unrecognized decomposed Clifford: {op!r}")
+        return self
+
+    def _imul_helper(self, other: 'cirq.PAULI_STRING_LIKE', sign: int):
+        """Left-multiplies or right-multiplies by a PAULI_STRING_LIKE.
+
+         Args:
+             other: What to multiply by.
+             sign: +1 to left-multiply, -1 to right-multiply.
+
+         Returns:
+             self on success, NotImplemented given an unknown type of value.
+        """
+        phase_log_i = 0
+
+        if isinstance(other, (Mapping, PauliString, MutablePauliString)):
+            if isinstance(other, (PauliString, MutablePauliString)):
+                self.coefficient *= other.coefficient
+            for qubit, pauli_gate_like in other.items():
+                pauli_int = _pauli_like_to_pauli_int(qubit, pauli_gate_like)
+                phase_log_i += self._imul_atom_helper(qubit, pauli_int, sign)
```

Comment:
I want to keep the amount of math involving complex numbers minimized. The code looks the way it does because that's how it's written in DensePauliString which uses this form because it can be done with a handful of numpy vector operations.

Fixed the numeric precision issue by switching to `self.coefficient *= 1j ** (phase_log_i & 3)`.

Reply:
Why though? I mean: this is a quantum computing library ;-) If there is any place for complex math anywhere in the world, then this is it! Also, in python it's built-in, so there is really no reason to avoid it.

Look, this does simplify the code. Compare to the alternative:

```python
for qubit, pauli_gate_like in other.items():
    pauli_int = _pauli_like_to_pauli_int(qubit, pauli_gate_like)
    self.coefficient *= self._imul_atom_helper(qubit, pauli_int, phase)
```

The present code is made more complicated by the presence of exponentiation, modulo arithmetic, hardcoded constant and unnecessary variable. And on top of that its intermediate values aren't as easy to interpret.

Reply:
Acknowledged, but I really do prefer this version.

Reply:
Adding `& 3` in the exponent to patch up the numerical issue leads to confusing code. One must consider subtleties of numerical exponentiation to appreciate why the `& 3` is even here. And since it appears unnecessary it's easy to imagine that it may be removed in future, inadvertently re-introducing numerical error.

Instead of patching it up, we should just remove the unnecessary complexity - the exponentiation, the bitmasking and the temporary variable `phase_log_i` - and instead we should have `_imul_atom_helper` return the phase correction to multiply directly into `self.coefficient` (rather than having it return one of the logarithms).

It leads to simpler, easily understandable code, doesn't it?

Reply:
I added a test that will fail if the `& 3` is removed.

I disagree that it's less clear. That being said, I am extremely comfortable with bit twiddling so it feels quite natural to me.

Adding ints is faster than multiplying complex numbers. Profiling:

```
    import cirq
    import time
    r = cirq.MutablePauliString()
    r2 = cirq.MutablePauliString()
    for i in range(1000):
        r[cirq.LineQubit(i)] = cirq.X
        r2[cirq.LineQubit(i)] = cirq.Y
    t0 = time.monotonic()
    for _ in range(1000):
        r *= r2
    t1 = time.monotonic()
    print(t1 - t0)
```

Using ints and bitmasks:

1.9

Using complex multiplication (factors made using 1j**sign):

2.1

Using complex multiplication (factors made using lookup from global R=[1,1j,-1,-1j] using R[sign]):

2.1

(There's some variation, but the ints are consistently lower than the complex.)

Reply:
It's a minor performance improvement for what (in my mind) is unnecessarily complicated code, but fair enough - now at least I see one minor advantage to working in the logs.